{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9a2e7da1-3db7-4854-84c4-c8167f04381d",
   "metadata": {},
   "source": [
    "# This converts relative coordinates (from LLM output) to actual genome coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bc3ff52e-a6bb-4436-b61c-62a83e335a36",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:02:45.654333Z",
     "iopub.status.busy": "2024-12-04T21:02:45.653771Z",
     "iopub.status.idle": "2024-12-04T21:02:48.105269Z",
     "shell.execute_reply": "2024-12-04T21:02:48.103830Z",
     "shell.execute_reply.started": "2024-12-04T21:02:45.654307Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import glob\n",
    "import gffutils\n",
    "from collections import defaultdict\n",
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "730d0654-56d8-4a78-aa06-0c5235ffb431",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:02:48.117205Z",
     "iopub.status.busy": "2024-12-04T21:02:48.115899Z",
     "iopub.status.idle": "2024-12-04T21:02:48.122104Z",
     "shell.execute_reply": "2024-12-04T21:02:48.121557Z",
     "shell.execute_reply.started": "2024-12-04T21:02:48.117177Z"
    }
   },
   "outputs": [],
   "source": [
    "# input_file = 'inputs/40_nucleotide_deletion_differences_allseqs_names_low_threshold.csv'\n",
    "# input_file = 'inputs/deletion_effects_threshold_1_30_nucleotides.csv'\n",
    "# input_file = 'inputs/deletion_effects_effects_human_water_60_nucleotides.csv'\n",
    "input_file = 'inputs/deletion_effects_effects_human_water_30_nucleotides.csv'\n",
    "input_dir = 'inputs/'\n",
    "output_dir = 'outputs/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "752d98ed-0771-4e8b-b7c4-67df4dd2d792",
   "metadata": {},
   "source": [
    "## Use GTF to define transcript coordinates \n",
    "`regions_positions` contains exon coordinates for each transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9905698a-ddbe-4181-8ef6-20bec6f02cf0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:02:48.134448Z",
     "iopub.status.busy": "2024-12-04T21:02:48.132342Z",
     "iopub.status.idle": "2024-12-04T21:03:04.234872Z",
     "shell.execute_reply": "2024-12-04T21:03:04.234494Z",
     "shell.execute_reply.started": "2024-12-04T21:02:48.134422Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3417782/356886721.py:3: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gtf = pd.read_csv('/tscc/projects/ps-yeolab3/bay001/annotations/GRCh38/ensembl/release_110/Homo_sapiens.GRCh38.110.chr.gtf', sep='\\t', skiprows=5, names=gtf_header)\n",
      "/tmp/ipykernel_3417782/356886721.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  exons['name'] = exons['attribute'].apply(lambda x: x.split('transcript_id \\\"')[1].split('\\\";')[0])\n",
      "/tmp/ipykernel_3417782/356886721.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  exons['start0base'] = exons['start'] - 1\n",
      "/tmp/ipykernel_3417782/356886721.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  exons['score'] = '0'\n",
      "/tmp/ipykernel_3417782/356886721.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  exons['chrom'] = 'chr' + exons['seqname'].astype(str)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start0base</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182695</td>\n",
       "      <td>182746</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183131</td>\n",
       "      <td>183216</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183493</td>\n",
       "      <td>183571</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183739</td>\n",
       "      <td>183901</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183980</td>\n",
       "      <td>184174</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  start0base     end             name score strand\n",
       "2  chr1      182695  182746  ENST00000624431     0      +\n",
       "3  chr1      183131  183216  ENST00000624431     0      +\n",
       "4  chr1      183493  183571  ENST00000624431     0      +\n",
       "5  chr1      183739  183901  ENST00000624431     0      +\n",
       "6  chr1      183980  184174  ENST00000624431     0      +"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_header = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "# gtf = pd.read_csv('inputs/Mus_musculus.GRCm39.112.chr.gtf', sep='\\t', skiprows=5, names=gtf_header)\n",
    "gtf = pd.read_csv('inputs/Homo_sapiens.GRCh38.110.chr.gtf', sep='\\t', skiprows=5, names=gtf_header)\n",
    "exons = gtf[gtf['feature']=='exon']\n",
    "exons['name'] = exons['attribute'].apply(lambda x: x.split('transcript_id \\\"')[1].split('\\\";')[0])\n",
    "\n",
    "exons.head()\n",
    "exons['start0base'] = exons['start'] - 1\n",
    "exons['score'] = '0'\n",
    "exons['chrom'] = 'chr' + exons['seqname'].astype(str)\n",
    "exons = exons[['chrom', 'start0base', 'end', 'name', 'score', 'strand']]\n",
    "exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2930011b-7ba8-47a0-88c4-8148e1fd8af4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:03:04.239538Z",
     "iopub.status.busy": "2024-12-04T21:03:04.238726Z",
     "iopub.status.idle": "2024-12-04T21:03:05.721210Z",
     "shell.execute_reply": "2024-12-04T21:03:05.720875Z",
     "shell.execute_reply.started": "2024-12-04T21:03:04.239520Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3417782/4141969072.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  transcripts['id'] = transcripts['attribute'].apply(lambda x: x.split('transcript_id \\\"')[1].split('\\\";')[0])\n",
      "/tmp/ipykernel_3417782/4141969072.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  transcripts['name'] = transcripts.apply(get_transcript_name, axis=1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>182696</td>\n",
       "      <td>184174</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000279928\"; gene_version \"2\"; transcript_id \"ENST00000624431\"; transcript_version \"2\"; gene_name \"DDX11L17\"; gene_source \"havana\"; gene_biotype \"unprocessed_pseudogene\"; transcript_name \"DDX11L17-201\"; transcript_source \"havana\"; transcript_biotype \"unprocessed_pseudogene\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"NA\";</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>DDX11L17-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2581560</td>\n",
       "      <td>2584533</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000228037\"; gene_version \"1\"; transcript_id \"ENST00000424215\"; transcript_version \"1\"; gene_source \"havana\"; gene_biotype \"lncRNA\"; transcript_source \"havana\"; transcript_biotype \"lncRNA\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"5\";</td>\n",
       "      <td>ENST00000424215</td>\n",
       "      <td>ENST00000424215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3434342</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; transcript_id \"ENST00000511072\"; transcript_version \"5\"; gene_name \"PRDM16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"PRDM16-206\"; transcript_source \"havana\"; transcript_biotype \"protein_coding\"; tag \"basic\"; transcript_support_level \"5\";</td>\n",
       "      <td>ENST00000511072</td>\n",
       "      <td>PRDM16-206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3069183</td>\n",
       "      <td>3186591</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; transcript_id \"ENST00000607632\"; transcript_version \"1\"; gene_name \"PRDM16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"PRDM16-210\"; transcript_source \"havana\"; transcript_biotype \"retained_intron\"; transcript_support_level \"2\";</td>\n",
       "      <td>ENST00000607632</td>\n",
       "      <td>PRDM16-210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3069197</td>\n",
       "      <td>3435421</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; transcript_id \"ENST00000378391\"; transcript_version \"6\"; gene_name \"PRDM16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"PRDM16-203\"; transcript_source \"havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS44048\"; tag \"basic\"; transcript_support_level \"1\";</td>\n",
       "      <td>ENST00000378391</td>\n",
       "      <td>PRDM16-203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3421027</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198695\"; gene_version \"2\"; transcript_id \"ENST00000361681\"; transcript_version \"2\"; gene_name \"MT-ND6\"; gene_source \"insdc\"; gene_biotype \"protein_coding\"; transcript_name \"MT-ND6-201\"; transcript_source \"insdc\"; transcript_biotype \"protein_coding\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"NA\";</td>\n",
       "      <td>ENST00000361681</td>\n",
       "      <td>MT-ND6-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3421033</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210194\"; gene_version \"1\"; transcript_id \"ENST00000387459\"; transcript_version \"1\"; gene_name \"MT-TE\"; gene_source \"insdc\"; gene_biotype \"Mt_tRNA\"; transcript_name \"MT-TE-201\"; transcript_source \"insdc\"; transcript_biotype \"Mt_tRNA\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"NA\";</td>\n",
       "      <td>ENST00000387459</td>\n",
       "      <td>MT-TE-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3421036</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198727\"; gene_version \"2\"; transcript_id \"ENST00000361789\"; transcript_version \"2\"; gene_name \"MT-CYB\"; gene_source \"insdc\"; gene_biotype \"protein_coding\"; transcript_name \"MT-CYB-201\"; transcript_source \"insdc\"; transcript_biotype \"protein_coding\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"NA\";</td>\n",
       "      <td>ENST00000361789</td>\n",
       "      <td>MT-CYB-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3421041</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195\"; gene_version \"2\"; transcript_id \"ENST00000387460\"; transcript_version \"2\"; gene_name \"MT-TT\"; gene_source \"insdc\"; gene_biotype \"Mt_tRNA\"; transcript_name \"MT-TT-201\"; transcript_source \"insdc\"; transcript_biotype \"Mt_tRNA\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"NA\";</td>\n",
       "      <td>ENST00000387460</td>\n",
       "      <td>MT-TT-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3421044</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196\"; gene_version \"2\"; transcript_id \"ENST00000387461\"; transcript_version \"2\"; gene_name \"MT-TP\"; gene_source \"insdc\"; gene_biotype \"Mt_tRNA\"; transcript_name \"MT-TP-201\"; transcript_source \"insdc\"; transcript_biotype \"Mt_tRNA\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"NA\";</td>\n",
       "      <td>ENST00000387461</td>\n",
       "      <td>MT-TP-201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>252835 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname  source     feature    start      end score strand frame  \\\n",
       "1             1  havana  transcript   182696   184174     .      +     .   \n",
       "8             1  havana  transcript  2581560  2584533     .      +     .   \n",
       "13            1  havana  transcript  3069168  3434342     .      +     .   \n",
       "50            1  havana  transcript  3069183  3186591     .      +     .   \n",
       "53            1  havana  transcript  3069197  3435421     .      +     .   \n",
       "...         ...     ...         ...      ...      ...   ...    ...   ...   \n",
       "3421027      MT   insdc  transcript    14149    14673     .      -     .   \n",
       "3421033      MT   insdc  transcript    14674    14742     .      -     .   \n",
       "3421036      MT   insdc  transcript    14747    15887     .      +     .   \n",
       "3421041      MT   insdc  transcript    15888    15953     .      +     .   \n",
       "3421044      MT   insdc  transcript    15956    16023     .      -     .   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                      attribute  \\\n",
       "1        gene_id \"ENSG00000279928\"; gene_version \"2\"; transcript_id \"ENST00000624431\"; transcript_version \"2\"; gene_name \"DDX11L17\"; gene_source \"havana\"; gene_biotype \"unprocessed_pseudogene\"; transcript_name \"DDX11L17-201\"; transcript_source \"havana\"; transcript_biotype \"unprocessed_pseudogene\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"NA\";   \n",
       "8                                                                                               gene_id \"ENSG00000228037\"; gene_version \"1\"; transcript_id \"ENST00000424215\"; transcript_version \"1\"; gene_source \"havana\"; gene_biotype \"lncRNA\"; transcript_source \"havana\"; transcript_biotype \"lncRNA\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"5\";   \n",
       "13                                            gene_id \"ENSG00000142611\"; gene_version \"17\"; transcript_id \"ENST00000511072\"; transcript_version \"5\"; gene_name \"PRDM16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"PRDM16-206\"; transcript_source \"havana\"; transcript_biotype \"protein_coding\"; tag \"basic\"; transcript_support_level \"5\";   \n",
       "50                                                        gene_id \"ENSG00000142611\"; gene_version \"17\"; transcript_id \"ENST00000607632\"; transcript_version \"1\"; gene_name \"PRDM16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"PRDM16-210\"; transcript_source \"havana\"; transcript_biotype \"retained_intron\"; transcript_support_level \"2\";   \n",
       "53           gene_id \"ENSG00000142611\"; gene_version \"17\"; transcript_id \"ENST00000378391\"; transcript_version \"6\"; gene_name \"PRDM16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"PRDM16-203\"; transcript_source \"havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS44048\"; tag \"basic\"; transcript_support_level \"1\";   \n",
       "...                                                                                                                                                                                                                                                                                                                                                                         ...   \n",
       "3421027                        gene_id \"ENSG00000198695\"; gene_version \"2\"; transcript_id \"ENST00000361681\"; transcript_version \"2\"; gene_name \"MT-ND6\"; gene_source \"insdc\"; gene_biotype \"protein_coding\"; transcript_name \"MT-ND6-201\"; transcript_source \"insdc\"; transcript_biotype \"protein_coding\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"NA\";   \n",
       "3421033                                        gene_id \"ENSG00000210194\"; gene_version \"1\"; transcript_id \"ENST00000387459\"; transcript_version \"1\"; gene_name \"MT-TE\"; gene_source \"insdc\"; gene_biotype \"Mt_tRNA\"; transcript_name \"MT-TE-201\"; transcript_source \"insdc\"; transcript_biotype \"Mt_tRNA\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"NA\";   \n",
       "3421036                        gene_id \"ENSG00000198727\"; gene_version \"2\"; transcript_id \"ENST00000361789\"; transcript_version \"2\"; gene_name \"MT-CYB\"; gene_source \"insdc\"; gene_biotype \"protein_coding\"; transcript_name \"MT-CYB-201\"; transcript_source \"insdc\"; transcript_biotype \"protein_coding\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"NA\";   \n",
       "3421041                                        gene_id \"ENSG00000210195\"; gene_version \"2\"; transcript_id \"ENST00000387460\"; transcript_version \"2\"; gene_name \"MT-TT\"; gene_source \"insdc\"; gene_biotype \"Mt_tRNA\"; transcript_name \"MT-TT-201\"; transcript_source \"insdc\"; transcript_biotype \"Mt_tRNA\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"NA\";   \n",
       "3421044                                        gene_id \"ENSG00000210196\"; gene_version \"2\"; transcript_id \"ENST00000387461\"; transcript_version \"2\"; gene_name \"MT-TP\"; gene_source \"insdc\"; gene_biotype \"Mt_tRNA\"; transcript_name \"MT-TP-201\"; transcript_source \"insdc\"; transcript_biotype \"Mt_tRNA\"; tag \"basic\"; tag \"Ensembl_canonical\"; transcript_support_level \"NA\";   \n",
       "\n",
       "                      id             name  \n",
       "1        ENST00000624431     DDX11L17-201  \n",
       "8        ENST00000424215  ENST00000424215  \n",
       "13       ENST00000511072       PRDM16-206  \n",
       "50       ENST00000607632       PRDM16-210  \n",
       "53       ENST00000378391       PRDM16-203  \n",
       "...                  ...              ...  \n",
       "3421027  ENST00000361681       MT-ND6-201  \n",
       "3421033  ENST00000387459        MT-TE-201  \n",
       "3421036  ENST00000361789       MT-CYB-201  \n",
       "3421041  ENST00000387460        MT-TT-201  \n",
       "3421044  ENST00000387461        MT-TP-201  \n",
       "\n",
       "[252835 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts = gtf[gtf['feature']=='transcript']\n",
    "\n",
    "def get_transcript_name(row):\n",
    "    try:\n",
    "        return row['attribute'].split('transcript_name \\\"')[1].split('\\\";')[0]\n",
    "    except Exception as e:\n",
    "        return row['attribute'].split('transcript_id \\\"')[1].split('\\\";')[0]\n",
    "\n",
    "transcripts['id'] = transcripts['attribute'].apply(lambda x: x.split('transcript_id \\\"')[1].split('\\\";')[0])\n",
    "transcripts['name'] = transcripts.apply(get_transcript_name, axis=1)\n",
    "transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0259bd14-16fe-4e13-91b8-6a06104755f0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:03:05.723447Z",
     "iopub.status.busy": "2024-12-04T21:03:05.722600Z",
     "iopub.status.idle": "2024-12-04T21:03:05.875249Z",
     "shell.execute_reply": "2024-12-04T21:03:05.874520Z",
     "shell.execute_reply.started": "2024-12-04T21:03:05.723427Z"
    }
   },
   "outputs": [],
   "source": [
    "tx2name = transcripts[['id', 'name']].set_index('id')['name'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "003c20fa-8856-4792-8246-e551d86c7fd0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:03:05.883237Z",
     "iopub.status.busy": "2024-12-04T21:03:05.882417Z",
     "iopub.status.idle": "2024-12-04T21:03:05.898047Z",
     "shell.execute_reply": "2024-12-04T21:03:05.896816Z",
     "shell.execute_reply.started": "2024-12-04T21:03:05.883220Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start0base</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182695</td>\n",
       "      <td>182746</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183131</td>\n",
       "      <td>183216</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183493</td>\n",
       "      <td>183571</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183739</td>\n",
       "      <td>183901</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183980</td>\n",
       "      <td>184174</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3421028</th>\n",
       "      <td>chrMT</td>\n",
       "      <td>14148</td>\n",
       "      <td>14673</td>\n",
       "      <td>ENST00000361681</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3421034</th>\n",
       "      <td>chrMT</td>\n",
       "      <td>14673</td>\n",
       "      <td>14742</td>\n",
       "      <td>ENST00000387459</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3421037</th>\n",
       "      <td>chrMT</td>\n",
       "      <td>14746</td>\n",
       "      <td>15887</td>\n",
       "      <td>ENST00000361789</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3421042</th>\n",
       "      <td>chrMT</td>\n",
       "      <td>15887</td>\n",
       "      <td>15953</td>\n",
       "      <td>ENST00000387460</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3421045</th>\n",
       "      <td>chrMT</td>\n",
       "      <td>15955</td>\n",
       "      <td>16023</td>\n",
       "      <td>ENST00000387461</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1649476 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chrom  start0base     end             name score strand\n",
       "2         chr1      182695  182746  ENST00000624431     0      +\n",
       "3         chr1      183131  183216  ENST00000624431     0      +\n",
       "4         chr1      183493  183571  ENST00000624431     0      +\n",
       "5         chr1      183739  183901  ENST00000624431     0      +\n",
       "6         chr1      183980  184174  ENST00000624431     0      +\n",
       "...        ...         ...     ...              ...   ...    ...\n",
       "3421028  chrMT       14148   14673  ENST00000361681     0      -\n",
       "3421034  chrMT       14673   14742  ENST00000387459     0      -\n",
       "3421037  chrMT       14746   15887  ENST00000361789     0      +\n",
       "3421042  chrMT       15887   15953  ENST00000387460     0      +\n",
       "3421045  chrMT       15955   16023  ENST00000387461     0      -\n",
       "\n",
       "[1649476 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ac712448-d17c-4a58-975d-179e08ef15b6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:03:05.900163Z",
     "iopub.status.busy": "2024-12-04T21:03:05.899346Z",
     "iopub.status.idle": "2024-12-04T21:03:05.917073Z",
     "shell.execute_reply": "2024-12-04T21:03:05.916488Z",
     "shell.execute_reply.started": "2024-12-04T21:03:05.900147Z"
    }
   },
   "outputs": [],
   "source": [
    "regions = exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6738f3cf-4707-44af-80a1-c90280223cdf",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:03:05.920844Z",
     "iopub.status.busy": "2024-12-04T21:03:05.919459Z",
     "iopub.status.idle": "2024-12-04T21:03:06.161922Z",
     "shell.execute_reply": "2024-12-04T21:03:06.161102Z",
     "shell.execute_reply.started": "2024-12-04T21:03:05.920817Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start0base</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182695</td>\n",
       "      <td>182746</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000624431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183131</td>\n",
       "      <td>183216</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000624431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183493</td>\n",
       "      <td>183571</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000624431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183739</td>\n",
       "      <td>183901</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000624431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183980</td>\n",
       "      <td>184174</td>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000624431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  start0base     end             name score strand    transcript_id\n",
       "2  chr1      182695  182746  ENST00000624431     0      +  ENST00000624431\n",
       "3  chr1      183131  183216  ENST00000624431     0      +  ENST00000624431\n",
       "4  chr1      183493  183571  ENST00000624431     0      +  ENST00000624431\n",
       "5  chr1      183739  183901  ENST00000624431     0      +  ENST00000624431\n",
       "6  chr1      183980  184174  ENST00000624431     0      +  ENST00000624431"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions['transcript_id'] = regions['name'].apply(lambda x: x.split('.')[0])  # in case we use gencode\n",
    "regions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b1bf9dbf-292f-4a97-8d16-69ac580d3b8a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:03:06.164184Z",
     "iopub.status.busy": "2024-12-04T21:03:06.163347Z",
     "iopub.status.idle": "2024-12-04T21:04:09.197945Z",
     "shell.execute_reply": "2024-12-04T21:04:09.197216Z",
     "shell.execute_reply.started": "2024-12-04T21:03:06.164166Z"
    }
   },
   "outputs": [],
   "source": [
    "regions.sort_values(by=['transcript_id','start0base','end'], inplace=True)  # critical for appending in order.\n",
    "regions_positions = defaultdict(list)\n",
    "for row in regions.iterrows():\n",
    "    regions_positions[row[1]['transcript_id']].append({'chrom': row[1]['chrom'], 'start': row[1]['start0base'], 'end': row[1]['end'], 'strand': row[1]['strand']})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ad574a32-b2df-46fc-b998-3e8c3f2e4564",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:04:09.200581Z",
     "iopub.status.busy": "2024-12-04T21:04:09.199682Z",
     "iopub.status.idle": "2024-12-04T21:04:09.204617Z",
     "shell.execute_reply": "2024-12-04T21:04:09.204301Z",
     "shell.execute_reply.started": "2024-12-04T21:04:09.200562Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'chrom': 'chr1', 'start': 182695, 'end': 182746, 'strand': '+'},\n",
       " {'chrom': 'chr1', 'start': 183131, 'end': 183216, 'strand': '+'},\n",
       " {'chrom': 'chr1', 'start': 183493, 'end': 183571, 'strand': '+'},\n",
       " {'chrom': 'chr1', 'start': 183739, 'end': 183901, 'strand': '+'},\n",
       " {'chrom': 'chr1', 'start': 183980, 'end': 184174, 'strand': '+'}]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# regions_positions['ENSMUST00000134384']\n",
    "regions_positions['ENST00000624431']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7dd61321-dd42-4bd7-8823-c1f489c6b159",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:04:09.206697Z",
     "iopub.status.busy": "2024-12-04T21:04:09.205893Z",
     "iopub.status.idle": "2024-12-04T21:04:36.469919Z",
     "shell.execute_reply": "2024-12-04T21:04:36.468878Z",
     "shell.execute_reply.started": "2024-12-04T21:04:09.206680Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of transcripts missing gene names: 4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>deletion_index_start</th>\n",
       "      <th>deletion_index_end</th>\n",
       "      <th>difference</th>\n",
       "      <th>synapse_probability</th>\n",
       "      <th>external_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044434</td>\n",
       "      <td>ARF5-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>15</td>\n",
       "      <td>45</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044434</td>\n",
       "      <td>ARF5-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>30</td>\n",
       "      <td>60</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044434</td>\n",
       "      <td>ARF5-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>45</td>\n",
       "      <td>75</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044434</td>\n",
       "      <td>ARF5-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>60</td>\n",
       "      <td>90</td>\n",
       "      <td>-0.001953</td>\n",
       "      <td>0.042480</td>\n",
       "      <td>ARF5-201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript_id  deletion_index_start  deletion_index_end  difference  \\\n",
       "0  ENST00000000233                     0                  30    0.000000   \n",
       "1  ENST00000000233                    15                  45    0.000000   \n",
       "2  ENST00000000233                    30                  60    0.000000   \n",
       "3  ENST00000000233                    45                  75    0.000000   \n",
       "4  ENST00000000233                    60                  90   -0.001953   \n",
       "\n",
       "   synapse_probability external_gene_name  \n",
       "0             0.044434           ARF5-201  \n",
       "1             0.044434           ARF5-201  \n",
       "2             0.044434           ARF5-201  \n",
       "3             0.044434           ARF5-201  \n",
       "4             0.042480           ARF5-201  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_name = set()\n",
    "def get_transcript_name(row, tx2name=tx2name):\n",
    "    try:\n",
    "        return tx2name[row['transcript_id']]\n",
    "    except KeyError:\n",
    "        missing_name.add(row['transcript_id'])\n",
    "        return row['transcript_id']\n",
    "        \n",
    "deletions = pd.read_csv(input_file)\n",
    "if 'external_gene_name' not in deletions.columns:\n",
    "    deletions['external_gene_name'] = deletions.apply(get_transcript_name, axis=1)\n",
    "\n",
    "print(f\"Number of transcripts missing gene names: {len(missing_name)}\")\n",
    "deletions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7d117db2-8581-4d78-a718-262940e533a1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:04:36.472402Z",
     "iopub.status.busy": "2024-12-04T21:04:36.471540Z",
     "iopub.status.idle": "2024-12-04T21:04:36.476150Z",
     "shell.execute_reply": "2024-12-04T21:04:36.475831Z",
     "shell.execute_reply.started": "2024-12-04T21:04:36.472382Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENST00000612465', 'ENST00000612640', 'ENST00000613204', 'ENST00000614336'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_name  # due to version mismatches in my GTF and Karen's. Should not be many but drop those that don't exist."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d3cc1cfd-9834-4d47-830f-e23dee27aa69",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:04:36.478297Z",
     "iopub.status.busy": "2024-12-04T21:04:36.477479Z",
     "iopub.status.idle": "2024-12-04T21:04:36.517281Z",
     "shell.execute_reply": "2024-12-04T21:04:36.516737Z",
     "shell.execute_reply.started": "2024-12-04T21:04:36.478280Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'chr1:183140-183180:+'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from itertools import groupby\n",
    "\n",
    "def collapse_ranges_groupby(data):\n",
    "    \"\"\"\n",
    "    Collapse sequences of numbers in a list into ranges using groupby.\n",
    "    \n",
    "    Args:\n",
    "      data: A list of integers.\n",
    "    \n",
    "    Returns:\n",
    "      A list of ranges represented as tuples (start, end).\n",
    "    \"\"\"\n",
    "    for _, group in groupby(enumerate(data), lambda pair: pair[1] - pair[0]):\n",
    "        group = list(group)\n",
    "        yield f\"{group[0][1]}-{group[-1][1]+1}\"\n",
    "      \n",
    "def get_gene_coords_from_tx_id(tx_id, relstart, relend, pos=regions_positions):\n",
    "\n",
    "    intervals = pos[tx_id]\n",
    "    if len(intervals) == 0: return \"\"  # utr not found\n",
    "    chrom = intervals[0]['chrom']\n",
    "    strand = list(set([s['strand'] for s in intervals]))\n",
    "    try:\n",
    "        assert len(strand) == 1\n",
    "        strand = strand[0]\n",
    "    except AssertionError:  # One transcript can't be associated with multiple strands?\n",
    "        # print(f\"{tx_id} exons are associated with more than one strand: {intervals}\")\n",
    "        return \"\"\n",
    "    # print(intervals)\n",
    "    utr_coords = [i for u in intervals for i in range(u['start'], u['end'])]\n",
    "    # print(utr_coords)\n",
    "    if len(utr_coords) < relend-relstart: return \"\"\n",
    "    if strand == '+':\n",
    "        true_coords = utr_coords[relstart:relend] \n",
    "        # print(true_coords)\n",
    "    elif strand == '-':\n",
    "        if relstart == 0:  # -0 breaks lists\n",
    "            true_coords = utr_coords[-relend:]\n",
    "        else:\n",
    "            true_coords = utr_coords[-relend:-relstart]\n",
    "    return ','.join([f\"{chrom}:{pos}:{strand}\" for pos in list(collapse_ranges_groupby(true_coords))])\n",
    "\n",
    "get_gene_coords_from_tx_id(tx_id='ENST00000624431', relstart=60, relend=100, pos=regions_positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e80caad-38e3-4122-87f3-bf72b46e86ba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "36a53dc0-2264-473d-ad42-56164fbd8c55",
   "metadata": {},
   "source": [
    "### Some unit tests to ensure code is correct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "484dbe26-9d0b-4dd5-ac7d-c887bd8c8de1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:04:36.520735Z",
     "iopub.status.busy": "2024-12-04T21:04:36.519656Z",
     "iopub.status.idle": "2024-12-04T21:04:36.563117Z",
     "shell.execute_reply": "2024-12-04T21:04:36.562310Z",
     "shell.execute_reply.started": "2024-12-04T21:04:36.520719Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1:1040-1080:+,chr1:2000-2001:+\n",
      "chr1:3010-3050:+\n",
      "chr1:2005-2010:+,chr1:3000-3014:+\n",
      "chr1:2019-2060:-\n",
      "chr1:999-1020:-,chr1:2000-2010:-,chr1:3000-3010:-\n",
      "chr1:3006-3025:-\n"
     ]
    }
   ],
   "source": [
    "test_positions = {}\n",
    "test_positions['test_pos_1'] = [\n",
    "    {'chrom': 'chr1', 'start': 1000, 'end': 1080, 'strand': '+'},\n",
    "    {'chrom': 'chr1', 'start': 2000, 'end': 2100, 'strand': '+'}\n",
    "]\n",
    "test_positions['test_pos_2'] = [\n",
    "    {'chrom': 'chr1', 'start': 1000, 'end': 1020, 'strand': '+'},\n",
    "    {'chrom': 'chr1', 'start': 2000, 'end': 2010, 'strand': '+'},\n",
    "    {'chrom': 'chr1', 'start': 3000, 'end': 3050, 'strand': '+'},\n",
    "]\n",
    "test_positions['test_pos_3'] = [\n",
    "    {'chrom': 'chr1', 'start': 1000, 'end': 1020, 'strand': '+'},\n",
    "    {'chrom': 'chr1', 'start': 2000, 'end': 2010, 'strand': '+'},\n",
    "    {'chrom': 'chr1', 'start': 3000, 'end': 3050, 'strand': '+'},\n",
    "]\n",
    "test_positions['test_pos_4'] = [\n",
    "    {'chrom': 'chr1', 'start': 1000, 'end': 1080, 'strand': '-'},\n",
    "    {'chrom': 'chr1', 'start': 2000, 'end': 2100, 'strand': '-'}\n",
    "]\n",
    "test_positions['test_pos_5'] = [\n",
    "    {'chrom': 'chr1', 'start': 0, 'end': 1020, 'strand': '-'},\n",
    "    {'chrom': 'chr1', 'start': 2000, 'end': 2010, 'strand': '-'},\n",
    "    {'chrom': 'chr1', 'start': 3000, 'end': 3050, 'strand': '-'},\n",
    "]\n",
    "test_positions['test_pos_6'] = [\n",
    "    {'chrom': 'chr1', 'start': 1000, 'end': 1020, 'strand': '-'},\n",
    "    {'chrom': 'chr1', 'start': 2000, 'end': 2010, 'strand': '-'},\n",
    "    {'chrom': 'chr1', 'start': 3000, 'end': 3050, 'strand': '-'},\n",
    "]\n",
    "print(get_gene_coords_from_tx_id(tx_id='test_pos_1', relstart=40, relend=81, pos=test_positions)) # chr1:1040-1080:+,chr1:2000-2001:+\n",
    "print(get_gene_coords_from_tx_id(tx_id='test_pos_2', relstart=40, relend=81, pos=test_positions)) # chr1:3010-3050:+\n",
    "print(get_gene_coords_from_tx_id(tx_id='test_pos_3', relstart=25, relend=44, pos=test_positions)) # chr1:2005-2010:+,chr1:3000-3014:+\n",
    "\n",
    "print(get_gene_coords_from_tx_id(tx_id='test_pos_4', relstart=40, relend=81, pos=test_positions)) # chr1:2019-2060:-\n",
    "print(get_gene_coords_from_tx_id(tx_id='test_pos_5', relstart=40, relend=81, pos=test_positions)) # chr1:999-1020:-,chr1:2000-2010:-,chr1:3000-3010:-\n",
    "print(get_gene_coords_from_tx_id(tx_id='test_pos_6', relstart=25, relend=44, pos=test_positions)) # chr1:3006-3025:-"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1c40c49-5018-4515-bb67-e986c5d93a49",
   "metadata": {},
   "source": [
    "## Iterate over each row in deletion and return true coordinates from relative ones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95d0bf4d-fd4b-48ef-8b97-0b0de5bbfbd7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:04:36.566275Z",
     "iopub.status.busy": "2024-12-04T21:04:36.565041Z"
    }
   },
   "outputs": [],
   "source": [
    "genomic_coords = {}\n",
    "for row in deletions.iterrows():\n",
    "    genomic_coords[row[0]] = get_gene_coords_from_tx_id(tx_id=row[1]['transcript_id'], relstart=row[1]['deletion_index_start'], relend=row[1]['deletion_index_end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "50906db1-30d1-46f7-a6ba-758ac228278a",
   "metadata": {
    "execution": {
     "iopub.status.idle": "2024-12-04T21:23:37.505233Z",
     "shell.execute_reply": "2024-12-04T21:23:37.504109Z",
     "shell.execute_reply.started": "2024-12-04T21:22:56.576303Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coords</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7:127588410-127588440:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7:127588425-127588455:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7:127588440-127588470:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7:127588455-127588485:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7:127588470-127588500:+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       coords\n",
       "0  chr7:127588410-127588440:+\n",
       "1  chr7:127588425-127588455:+\n",
       "2  chr7:127588440-127588470:+\n",
       "3  chr7:127588455-127588485:+\n",
       "4  chr7:127588470-127588500:+"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomic_coords = pd.DataFrame(genomic_coords, index=['coords']).T\n",
    "genomic_coords.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9387e20-2f8d-467c-a2b8-b1ac583d14fa",
   "metadata": {},
   "source": [
    "## Merge true coordinates with deletions table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f57eb7e9-1c81-4bdc-8799-397cb2e64c31",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:23:37.507325Z",
     "iopub.status.busy": "2024-12-04T21:23:37.506548Z",
     "iopub.status.idle": "2024-12-04T21:23:38.170783Z",
     "shell.execute_reply": "2024-12-04T21:23:38.169584Z",
     "shell.execute_reply.started": "2024-12-04T21:23:37.507307Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>deletion_index_start</th>\n",
       "      <th>deletion_index_end</th>\n",
       "      <th>difference</th>\n",
       "      <th>synapse_probability</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>coords</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044434</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127588410-127588440:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>15</td>\n",
       "      <td>45</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044434</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127588425-127588455:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>30</td>\n",
       "      <td>60</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044434</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127588440-127588470:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>45</td>\n",
       "      <td>75</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044434</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127588455-127588485:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>60</td>\n",
       "      <td>90</td>\n",
       "      <td>-0.001953</td>\n",
       "      <td>0.042480</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127588470-127588500:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6969881</th>\n",
       "      <td>ENST00000707136</td>\n",
       "      <td>2835</td>\n",
       "      <td>2865</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018433</td>\n",
       "      <td>PRRT3-205</td>\n",
       "      <td>chr3:9939588-9939618:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6969882</th>\n",
       "      <td>ENST00000707136</td>\n",
       "      <td>2850</td>\n",
       "      <td>2880</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018433</td>\n",
       "      <td>PRRT3-205</td>\n",
       "      <td>chr3:9939573-9939603:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6969883</th>\n",
       "      <td>ENST00000707136</td>\n",
       "      <td>2865</td>\n",
       "      <td>2895</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018433</td>\n",
       "      <td>PRRT3-205</td>\n",
       "      <td>chr3:9939558-9939588:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6969884</th>\n",
       "      <td>ENST00000707136</td>\n",
       "      <td>2880</td>\n",
       "      <td>2910</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018433</td>\n",
       "      <td>PRRT3-205</td>\n",
       "      <td>chr3:9939543-9939573:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6969885</th>\n",
       "      <td>ENST00000707136</td>\n",
       "      <td>2895</td>\n",
       "      <td>2925</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018433</td>\n",
       "      <td>PRRT3-205</td>\n",
       "      <td>chr3:9939528-9939558:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6969886 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           transcript_id  deletion_index_start  deletion_index_end  \\\n",
       "0        ENST00000000233                     0                  30   \n",
       "1        ENST00000000233                    15                  45   \n",
       "2        ENST00000000233                    30                  60   \n",
       "3        ENST00000000233                    45                  75   \n",
       "4        ENST00000000233                    60                  90   \n",
       "...                  ...                   ...                 ...   \n",
       "6969881  ENST00000707136                  2835                2865   \n",
       "6969882  ENST00000707136                  2850                2880   \n",
       "6969883  ENST00000707136                  2865                2895   \n",
       "6969884  ENST00000707136                  2880                2910   \n",
       "6969885  ENST00000707136                  2895                2925   \n",
       "\n",
       "         difference  synapse_probability external_gene_name  \\\n",
       "0          0.000000             0.044434           ARF5-201   \n",
       "1          0.000000             0.044434           ARF5-201   \n",
       "2          0.000000             0.044434           ARF5-201   \n",
       "3          0.000000             0.044434           ARF5-201   \n",
       "4         -0.001953             0.042480           ARF5-201   \n",
       "...             ...                  ...                ...   \n",
       "6969881    0.000000             0.018433          PRRT3-205   \n",
       "6969882    0.000000             0.018433          PRRT3-205   \n",
       "6969883    0.000000             0.018433          PRRT3-205   \n",
       "6969884    0.000000             0.018433          PRRT3-205   \n",
       "6969885    0.000000             0.018433          PRRT3-205   \n",
       "\n",
       "                             coords  \n",
       "0        chr7:127588410-127588440:+  \n",
       "1        chr7:127588425-127588455:+  \n",
       "2        chr7:127588440-127588470:+  \n",
       "3        chr7:127588455-127588485:+  \n",
       "4        chr7:127588470-127588500:+  \n",
       "...                             ...  \n",
       "6969881      chr3:9939588-9939618:-  \n",
       "6969882      chr3:9939573-9939603:-  \n",
       "6969883      chr3:9939558-9939588:-  \n",
       "6969884      chr3:9939543-9939573:-  \n",
       "6969885      chr3:9939528-9939558:-  \n",
       "\n",
       "[6969886 rows x 7 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = pd.merge(deletions, genomic_coords, how='outer', left_index=True, right_index=True)\n",
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "128768e1-d13e-41e4-8b05-4a5e07bec74e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:23:38.172847Z",
     "iopub.status.busy": "2024-12-04T21:23:38.172074Z",
     "iopub.status.idle": "2024-12-04T21:23:39.514919Z",
     "shell.execute_reply": "2024-12-04T21:23:39.514570Z",
     "shell.execute_reply.started": "2024-12-04T21:23:38.172829Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>deletion_index_start</th>\n",
       "      <th>deletion_index_end</th>\n",
       "      <th>difference</th>\n",
       "      <th>synapse_probability</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>coords</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>135</td>\n",
       "      <td>165</td>\n",
       "      <td>0.000977</td>\n",
       "      <td>0.045410</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127588545-127588565:+,chr7:127589082-127589092:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>150</td>\n",
       "      <td>180</td>\n",
       "      <td>0.014893</td>\n",
       "      <td>0.059326</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127588560-127588565:+,chr7:127589082-127589107:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>210</td>\n",
       "      <td>240</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044434</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127589137-127589163:+,chr7:127589484-127589488:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>225</td>\n",
       "      <td>255</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044434</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127589152-127589163:+,chr7:127589484-127589503:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>330</td>\n",
       "      <td>360</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044434</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127589578-127589594:+,chr7:127590065-127590079:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>345</td>\n",
       "      <td>375</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044434</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127589593-127589594:+,chr7:127590065-127590094:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>390</td>\n",
       "      <td>420</td>\n",
       "      <td>0.001709</td>\n",
       "      <td>0.046143</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127590109-127590137:+,chr7:127590962-127590964:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>405</td>\n",
       "      <td>435</td>\n",
       "      <td>0.001953</td>\n",
       "      <td>0.046387</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127590124-127590137:+,chr7:127590962-127590979:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>525</td>\n",
       "      <td>555</td>\n",
       "      <td>0.001221</td>\n",
       "      <td>0.045654</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127591069-127591088:+,chr7:127591212-127591223:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>540</td>\n",
       "      <td>570</td>\n",
       "      <td>0.001221</td>\n",
       "      <td>0.045654</td>\n",
       "      <td>ARF5-201</td>\n",
       "      <td>chr7:127591084-127591088:+,chr7:127591212-127591238:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>135</td>\n",
       "      <td>165</td>\n",
       "      <td>-0.189453</td>\n",
       "      <td>0.078125</td>\n",
       "      <td>M6PR-201</td>\n",
       "      <td>chr12:8946398-8946405:-,chr12:8949487-8949510:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>150</td>\n",
       "      <td>180</td>\n",
       "      <td>-0.152344</td>\n",
       "      <td>0.115234</td>\n",
       "      <td>M6PR-201</td>\n",
       "      <td>chr12:8946383-8946405:-,chr12:8949487-8949495:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>315</td>\n",
       "      <td>345</td>\n",
       "      <td>0.121094</td>\n",
       "      <td>0.388672</td>\n",
       "      <td>M6PR-201</td>\n",
       "      <td>chr12:8945574-8945584:-,chr12:8946228-8946248:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>330</td>\n",
       "      <td>360</td>\n",
       "      <td>0.193359</td>\n",
       "      <td>0.460938</td>\n",
       "      <td>M6PR-201</td>\n",
       "      <td>chr12:8945559-8945584:-,chr12:8946228-8946233:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>480</td>\n",
       "      <td>510</td>\n",
       "      <td>-0.049805</td>\n",
       "      <td>0.217773</td>\n",
       "      <td>M6PR-201</td>\n",
       "      <td>chr12:8943902-8943910:-,chr12:8945417-8945439:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>495</td>\n",
       "      <td>525</td>\n",
       "      <td>-0.059570</td>\n",
       "      <td>0.208008</td>\n",
       "      <td>M6PR-201</td>\n",
       "      <td>chr12:8943887-8943910:-,chr12:8945417-8945424:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>585</td>\n",
       "      <td>615</td>\n",
       "      <td>0.042969</td>\n",
       "      <td>0.310547</td>\n",
       "      <td>M6PR-201</td>\n",
       "      <td>chr12:8943532-8943535:-,chr12:8943800-8943827:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>600</td>\n",
       "      <td>630</td>\n",
       "      <td>0.023438</td>\n",
       "      <td>0.291016</td>\n",
       "      <td>M6PR-201</td>\n",
       "      <td>chr12:8943517-8943535:-,chr12:8943800-8943812:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>720</td>\n",
       "      <td>750</td>\n",
       "      <td>-0.093750</td>\n",
       "      <td>0.173828</td>\n",
       "      <td>M6PR-201</td>\n",
       "      <td>chr12:8942535-8942542:-,chr12:8943404-8943427:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>735</td>\n",
       "      <td>765</td>\n",
       "      <td>-0.114258</td>\n",
       "      <td>0.153320</td>\n",
       "      <td>M6PR-201</td>\n",
       "      <td>chr12:8942520-8942542:-,chr12:8943404-8943412:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>855</td>\n",
       "      <td>885</td>\n",
       "      <td>-0.011719</td>\n",
       "      <td>0.255859</td>\n",
       "      <td>M6PR-201</td>\n",
       "      <td>chr12:8941925-8941940:-,chr12:8942415-8942430:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>195</td>\n",
       "      <td>225</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>ESRRA-201</td>\n",
       "      <td>chr11:64305718-64305736:+,chr11:64307167-64307179:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>210</td>\n",
       "      <td>240</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>ESRRA-201</td>\n",
       "      <td>chr11:64305733-64305736:+,chr11:64307167-64307194:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>525</td>\n",
       "      <td>555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>ESRRA-201</td>\n",
       "      <td>chr11:64307479-64307504:+,chr11:64313950-64313955:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>540</td>\n",
       "      <td>570</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>ESRRA-201</td>\n",
       "      <td>chr11:64307494-64307504:+,chr11:64313950-64313970:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>645</td>\n",
       "      <td>675</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>ESRRA-201</td>\n",
       "      <td>chr11:64314045-64314067:+,chr11:64314238-64314246:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>660</td>\n",
       "      <td>690</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>ESRRA-201</td>\n",
       "      <td>chr11:64314060-64314067:+,chr11:64314238-64314261:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>780</td>\n",
       "      <td>810</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>ESRRA-201</td>\n",
       "      <td>chr11:64314351-64314367:+,chr11:64314740-64314754:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>795</td>\n",
       "      <td>825</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>ESRRA-201</td>\n",
       "      <td>chr11:64314366-64314367:+,chr11:64314740-64314769:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>945</td>\n",
       "      <td>975</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>ESRRA-201</td>\n",
       "      <td>chr11:64314889-64314911:+,chr11:64315000-64315008:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>960</td>\n",
       "      <td>990</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>ESRRA-201</td>\n",
       "      <td>chr11:64314904-64314911:+,chr11:64315000-64315023:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>1215</td>\n",
       "      <td>1245</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>ESRRA-201</td>\n",
       "      <td>chr11:64315248-64315270:+,chr11:64315706-64315714:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>1230</td>\n",
       "      <td>1260</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>ESRRA-201</td>\n",
       "      <td>chr11:64315263-64315270:+,chr11:64315706-64315729:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>255</td>\n",
       "      <td>285</td>\n",
       "      <td>0.000366</td>\n",
       "      <td>0.010681</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2795224-2795244:+,chr12:2797137-2797147:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>270</td>\n",
       "      <td>300</td>\n",
       "      <td>0.001038</td>\n",
       "      <td>0.011353</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2795239-2795244:+,chr12:2797137-2797162:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>405</td>\n",
       "      <td>435</td>\n",
       "      <td>0.000366</td>\n",
       "      <td>0.010681</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2797267-2797282:+,chr12:2797728-2797743:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>540</td>\n",
       "      <td>570</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010315</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2797848-2797871:+,chr12:2798705-2798712:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>555</td>\n",
       "      <td>585</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010315</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2797863-2797871:+,chr12:2798705-2798727:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>660</td>\n",
       "      <td>690</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010315</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2798802-2798826:+,chr12:2799087-2799093:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>675</td>\n",
       "      <td>705</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010315</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2798817-2798826:+,chr12:2799087-2799108:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>825</td>\n",
       "      <td>855</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010315</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2799228-2799244:+,chr12:2799849-2799863:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>840</td>\n",
       "      <td>870</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010315</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2799243-2799244:+,chr12:2799849-2799878:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>440</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>915</td>\n",
       "      <td>945</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>0.010498</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2799923-2799940:+,chr12:2800038-2800051:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>930</td>\n",
       "      <td>960</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>0.010498</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2799938-2799940:+,chr12:2800038-2800066:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>990</td>\n",
       "      <td>1020</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010315</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2800096-2800122:+,chr12:2800391-2800395:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>1005</td>\n",
       "      <td>1035</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>0.010498</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2800111-2800122:+,chr12:2800391-2800410:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>1185</td>\n",
       "      <td>1215</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>0.010498</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2800560-2800577:+,chr12:2801116-2801129:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>1200</td>\n",
       "      <td>1230</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010315</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2800575-2800577:+,chr12:2801116-2801144:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>1425</td>\n",
       "      <td>1455</td>\n",
       "      <td>0.000366</td>\n",
       "      <td>0.010681</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2801339-2801356:+,chr12:2803150-2803163:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>1440</td>\n",
       "      <td>1470</td>\n",
       "      <td>0.000366</td>\n",
       "      <td>0.010681</td>\n",
       "      <td>FKBP4-201</td>\n",
       "      <td>chr12:2801354-2801356:+,chr12:2803150-2803178:+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id  deletion_index_start  deletion_index_end  difference  \\\n",
       "9    ENST00000000233                   135                 165    0.000977   \n",
       "10   ENST00000000233                   150                 180    0.014893   \n",
       "14   ENST00000000233                   210                 240    0.000000   \n",
       "15   ENST00000000233                   225                 255    0.000000   \n",
       "22   ENST00000000233                   330                 360    0.000000   \n",
       "23   ENST00000000233                   345                 375    0.000000   \n",
       "26   ENST00000000233                   390                 420    0.001709   \n",
       "27   ENST00000000233                   405                 435    0.001953   \n",
       "35   ENST00000000233                   525                 555    0.001221   \n",
       "36   ENST00000000233                   540                 570    0.001221   \n",
       "76   ENST00000000412                   135                 165   -0.189453   \n",
       "77   ENST00000000412                   150                 180   -0.152344   \n",
       "88   ENST00000000412                   315                 345    0.121094   \n",
       "89   ENST00000000412                   330                 360    0.193359   \n",
       "99   ENST00000000412                   480                 510   -0.049805   \n",
       "100  ENST00000000412                   495                 525   -0.059570   \n",
       "106  ENST00000000412                   585                 615    0.042969   \n",
       "107  ENST00000000412                   600                 630    0.023438   \n",
       "115  ENST00000000412                   720                 750   -0.093750   \n",
       "116  ENST00000000412                   735                 765   -0.114258   \n",
       "124  ENST00000000412                   855                 885   -0.011719   \n",
       "242  ENST00000000442                   195                 225    0.000000   \n",
       "243  ENST00000000442                   210                 240    0.000000   \n",
       "264  ENST00000000442                   525                 555    0.000000   \n",
       "265  ENST00000000442                   540                 570    0.000000   \n",
       "272  ENST00000000442                   645                 675    0.000000   \n",
       "273  ENST00000000442                   660                 690    0.000000   \n",
       "281  ENST00000000442                   780                 810    0.000000   \n",
       "282  ENST00000000442                   795                 825    0.000000   \n",
       "292  ENST00000000442                   945                 975    0.000000   \n",
       "293  ENST00000000442                   960                 990    0.000000   \n",
       "310  ENST00000000442                  1215                1245    0.000000   \n",
       "311  ENST00000000442                  1230                1260    0.000000   \n",
       "396  ENST00000001008                   255                 285    0.000366   \n",
       "397  ENST00000001008                   270                 300    0.001038   \n",
       "406  ENST00000001008                   405                 435    0.000366   \n",
       "415  ENST00000001008                   540                 570    0.000000   \n",
       "416  ENST00000001008                   555                 585    0.000000   \n",
       "423  ENST00000001008                   660                 690    0.000000   \n",
       "424  ENST00000001008                   675                 705    0.000000   \n",
       "434  ENST00000001008                   825                 855    0.000000   \n",
       "435  ENST00000001008                   840                 870    0.000000   \n",
       "440  ENST00000001008                   915                 945    0.000183   \n",
       "441  ENST00000001008                   930                 960    0.000183   \n",
       "445  ENST00000001008                   990                1020    0.000000   \n",
       "446  ENST00000001008                  1005                1035    0.000183   \n",
       "458  ENST00000001008                  1185                1215    0.000183   \n",
       "459  ENST00000001008                  1200                1230    0.000000   \n",
       "474  ENST00000001008                  1425                1455    0.000366   \n",
       "475  ENST00000001008                  1440                1470    0.000366   \n",
       "\n",
       "     synapse_probability external_gene_name  \\\n",
       "9               0.045410           ARF5-201   \n",
       "10              0.059326           ARF5-201   \n",
       "14              0.044434           ARF5-201   \n",
       "15              0.044434           ARF5-201   \n",
       "22              0.044434           ARF5-201   \n",
       "23              0.044434           ARF5-201   \n",
       "26              0.046143           ARF5-201   \n",
       "27              0.046387           ARF5-201   \n",
       "35              0.045654           ARF5-201   \n",
       "36              0.045654           ARF5-201   \n",
       "76              0.078125           M6PR-201   \n",
       "77              0.115234           M6PR-201   \n",
       "88              0.388672           M6PR-201   \n",
       "89              0.460938           M6PR-201   \n",
       "99              0.217773           M6PR-201   \n",
       "100             0.208008           M6PR-201   \n",
       "106             0.310547           M6PR-201   \n",
       "107             0.291016           M6PR-201   \n",
       "115             0.173828           M6PR-201   \n",
       "116             0.153320           M6PR-201   \n",
       "124             0.255859           M6PR-201   \n",
       "242             0.013000          ESRRA-201   \n",
       "243             0.013000          ESRRA-201   \n",
       "264             0.013000          ESRRA-201   \n",
       "265             0.013000          ESRRA-201   \n",
       "272             0.013000          ESRRA-201   \n",
       "273             0.013000          ESRRA-201   \n",
       "281             0.013000          ESRRA-201   \n",
       "282             0.013000          ESRRA-201   \n",
       "292             0.013000          ESRRA-201   \n",
       "293             0.013000          ESRRA-201   \n",
       "310             0.013000          ESRRA-201   \n",
       "311             0.013000          ESRRA-201   \n",
       "396             0.010681          FKBP4-201   \n",
       "397             0.011353          FKBP4-201   \n",
       "406             0.010681          FKBP4-201   \n",
       "415             0.010315          FKBP4-201   \n",
       "416             0.010315          FKBP4-201   \n",
       "423             0.010315          FKBP4-201   \n",
       "424             0.010315          FKBP4-201   \n",
       "434             0.010315          FKBP4-201   \n",
       "435             0.010315          FKBP4-201   \n",
       "440             0.010498          FKBP4-201   \n",
       "441             0.010498          FKBP4-201   \n",
       "445             0.010315          FKBP4-201   \n",
       "446             0.010498          FKBP4-201   \n",
       "458             0.010498          FKBP4-201   \n",
       "459             0.010315          FKBP4-201   \n",
       "474             0.010681          FKBP4-201   \n",
       "475             0.010681          FKBP4-201   \n",
       "\n",
       "                                                    coords  \n",
       "9    chr7:127588545-127588565:+,chr7:127589082-127589092:+  \n",
       "10   chr7:127588560-127588565:+,chr7:127589082-127589107:+  \n",
       "14   chr7:127589137-127589163:+,chr7:127589484-127589488:+  \n",
       "15   chr7:127589152-127589163:+,chr7:127589484-127589503:+  \n",
       "22   chr7:127589578-127589594:+,chr7:127590065-127590079:+  \n",
       "23   chr7:127589593-127589594:+,chr7:127590065-127590094:+  \n",
       "26   chr7:127590109-127590137:+,chr7:127590962-127590964:+  \n",
       "27   chr7:127590124-127590137:+,chr7:127590962-127590979:+  \n",
       "35   chr7:127591069-127591088:+,chr7:127591212-127591223:+  \n",
       "36   chr7:127591084-127591088:+,chr7:127591212-127591238:+  \n",
       "76         chr12:8946398-8946405:-,chr12:8949487-8949510:-  \n",
       "77         chr12:8946383-8946405:-,chr12:8949487-8949495:-  \n",
       "88         chr12:8945574-8945584:-,chr12:8946228-8946248:-  \n",
       "89         chr12:8945559-8945584:-,chr12:8946228-8946233:-  \n",
       "99         chr12:8943902-8943910:-,chr12:8945417-8945439:-  \n",
       "100        chr12:8943887-8943910:-,chr12:8945417-8945424:-  \n",
       "106        chr12:8943532-8943535:-,chr12:8943800-8943827:-  \n",
       "107        chr12:8943517-8943535:-,chr12:8943800-8943812:-  \n",
       "115        chr12:8942535-8942542:-,chr12:8943404-8943427:-  \n",
       "116        chr12:8942520-8942542:-,chr12:8943404-8943412:-  \n",
       "124        chr12:8941925-8941940:-,chr12:8942415-8942430:-  \n",
       "242    chr11:64305718-64305736:+,chr11:64307167-64307179:+  \n",
       "243    chr11:64305733-64305736:+,chr11:64307167-64307194:+  \n",
       "264    chr11:64307479-64307504:+,chr11:64313950-64313955:+  \n",
       "265    chr11:64307494-64307504:+,chr11:64313950-64313970:+  \n",
       "272    chr11:64314045-64314067:+,chr11:64314238-64314246:+  \n",
       "273    chr11:64314060-64314067:+,chr11:64314238-64314261:+  \n",
       "281    chr11:64314351-64314367:+,chr11:64314740-64314754:+  \n",
       "282    chr11:64314366-64314367:+,chr11:64314740-64314769:+  \n",
       "292    chr11:64314889-64314911:+,chr11:64315000-64315008:+  \n",
       "293    chr11:64314904-64314911:+,chr11:64315000-64315023:+  \n",
       "310    chr11:64315248-64315270:+,chr11:64315706-64315714:+  \n",
       "311    chr11:64315263-64315270:+,chr11:64315706-64315729:+  \n",
       "396        chr12:2795224-2795244:+,chr12:2797137-2797147:+  \n",
       "397        chr12:2795239-2795244:+,chr12:2797137-2797162:+  \n",
       "406        chr12:2797267-2797282:+,chr12:2797728-2797743:+  \n",
       "415        chr12:2797848-2797871:+,chr12:2798705-2798712:+  \n",
       "416        chr12:2797863-2797871:+,chr12:2798705-2798727:+  \n",
       "423        chr12:2798802-2798826:+,chr12:2799087-2799093:+  \n",
       "424        chr12:2798817-2798826:+,chr12:2799087-2799108:+  \n",
       "434        chr12:2799228-2799244:+,chr12:2799849-2799863:+  \n",
       "435        chr12:2799243-2799244:+,chr12:2799849-2799878:+  \n",
       "440        chr12:2799923-2799940:+,chr12:2800038-2800051:+  \n",
       "441        chr12:2799938-2799940:+,chr12:2800038-2800066:+  \n",
       "445        chr12:2800096-2800122:+,chr12:2800391-2800395:+  \n",
       "446        chr12:2800111-2800122:+,chr12:2800391-2800410:+  \n",
       "458        chr12:2800560-2800577:+,chr12:2801116-2801129:+  \n",
       "459        chr12:2800575-2800577:+,chr12:2801116-2801144:+  \n",
       "474        chr12:2801339-2801356:+,chr12:2803150-2803163:+  \n",
       "475        chr12:2801354-2801356:+,chr12:2803150-2803178:+  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Double check hotspots that span more than one exon\n",
    "merged[merged['coords'].str.contains(',')].head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0bdb5d41-00ed-4edc-8a18-be41518caef3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:23:39.516998Z",
     "iopub.status.busy": "2024-12-04T21:23:39.516227Z",
     "iopub.status.idle": "2024-12-04T21:23:53.182576Z",
     "shell.execute_reply": "2024-12-04T21:23:53.181508Z",
     "shell.execute_reply.started": "2024-12-04T21:23:39.516979Z"
    }
   },
   "outputs": [],
   "source": [
    "# merged.to_csv(os.path.join(output_dir, '40_nucleotide_deletion_differences_allseqs_names_low_threshold.coords.tsv'), sep='\\t', index=False, header=True)\n",
    "# merged.to_csv(os.path.join(output_dir, 'deletion_effects_threshold_1_30_nucleotides.coords.tsv'), sep='\\t', index=False, header=True)\n",
    "merged.to_csv(os.path.join(output_dir, os.path.basename(input_file).replace('.csv', '.coords.tsv')), sep='\\t', index=False, header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d78a73e6-7720-4e52-a3a6-2123b7eb64dd",
   "metadata": {},
   "source": [
    "## Check once more that regions spanning multiple exons are correct\n",
    "- Here we check regions that span at least three exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "15abe2e3-1a49-401f-8e30-07d0309bd844",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-04T21:23:53.188557Z",
     "iopub.status.busy": "2024-12-04T21:23:53.186347Z",
     "iopub.status.idle": "2024-12-04T21:23:54.885233Z",
     "shell.execute_reply": "2024-12-04T21:23:54.884510Z",
     "shell.execute_reply.started": "2024-12-04T21:23:53.188511Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>deletion_index_start</th>\n",
       "      <th>deletion_index_end</th>\n",
       "      <th>difference</th>\n",
       "      <th>synapse_probability</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>coords</th>\n",
       "      <th>num_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6271</th>\n",
       "      <td>ENST00000011619</td>\n",
       "      <td>1230</td>\n",
       "      <td>1260</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014038</td>\n",
       "      <td>RANBP9-201</td>\n",
       "      <td>chr6:13644726-13644729:-,chr6:13652658-13652681:-,chr6:13657108-13657112:-</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8725</th>\n",
       "      <td>ENST00000013125</td>\n",
       "      <td>2010</td>\n",
       "      <td>2040</td>\n",
       "      <td>0.000793</td>\n",
       "      <td>0.016235</td>\n",
       "      <td>MAP4K5-201</td>\n",
       "      <td>chr14:50437995-50438008:-,chr14:50438091-50438094:-,chr14:50440012-50440026:-</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19034</th>\n",
       "      <td>ENST00000064778</td>\n",
       "      <td>435</td>\n",
       "      <td>465</td>\n",
       "      <td>-0.000488</td>\n",
       "      <td>0.014038</td>\n",
       "      <td>FAM168A-201</td>\n",
       "      <td>chr11:73419997-73419999:-,chr11:73425021-73425048:-,chr11:73430689-73430690:-</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43148</th>\n",
       "      <td>ENST00000195649</td>\n",
       "      <td>1170</td>\n",
       "      <td>1200</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013428</td>\n",
       "      <td>SNAP91-201</td>\n",
       "      <td>chr6:83610675-83610677:-,chr6:83614856-83614862:-,chr6:83616968-83616990:-</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43149</th>\n",
       "      <td>ENST00000195649</td>\n",
       "      <td>1185</td>\n",
       "      <td>1215</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013428</td>\n",
       "      <td>SNAP91-201</td>\n",
       "      <td>chr6:83610660-83610677:-,chr6:83614856-83614862:-,chr6:83616968-83616975:-</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6964154</th>\n",
       "      <td>ENST00000706889</td>\n",
       "      <td>4710</td>\n",
       "      <td>4740</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.016357</td>\n",
       "      <td>MADD-227</td>\n",
       "      <td>chr11:47324565-47324577:+,chr11:47326543-47326552:+,chr11:47326737-47326746:+</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6965958</th>\n",
       "      <td>ENST00000706935</td>\n",
       "      <td>1500</td>\n",
       "      <td>1530</td>\n",
       "      <td>0.000305</td>\n",
       "      <td>0.009705</td>\n",
       "      <td>INPP4A-210</td>\n",
       "      <td>chr2:98546687-98546694:+,chr2:98548951-98548969:+,chr2:98552785-98552790:+</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6967922</th>\n",
       "      <td>ENST00000707119</td>\n",
       "      <td>1695</td>\n",
       "      <td>1725</td>\n",
       "      <td>-0.000244</td>\n",
       "      <td>0.013611</td>\n",
       "      <td>ABLIM1-215</td>\n",
       "      <td>chr10:114439245-114439250:-,chr10:114440081-114440089:-,chr10:114441016-114441033:-</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6967923</th>\n",
       "      <td>ENST00000707119</td>\n",
       "      <td>1710</td>\n",
       "      <td>1740</td>\n",
       "      <td>-0.000427</td>\n",
       "      <td>0.013428</td>\n",
       "      <td>ABLIM1-215</td>\n",
       "      <td>chr10:114439230-114439250:-,chr10:114440081-114440089:-,chr10:114441016-114441018:-</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6968522</th>\n",
       "      <td>ENST00000707127</td>\n",
       "      <td>735</td>\n",
       "      <td>765</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015076</td>\n",
       "      <td>MEF2C-257</td>\n",
       "      <td>chr5:88729344-88729347:-,chr5:88730210-88730234:-,chr5:88731728-88731731:-</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1167 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           transcript_id  deletion_index_start  deletion_index_end  \\\n",
       "6271     ENST00000011619                  1230                1260   \n",
       "8725     ENST00000013125                  2010                2040   \n",
       "19034    ENST00000064778                   435                 465   \n",
       "43148    ENST00000195649                  1170                1200   \n",
       "43149    ENST00000195649                  1185                1215   \n",
       "...                  ...                   ...                 ...   \n",
       "6964154  ENST00000706889                  4710                4740   \n",
       "6965958  ENST00000706935                  1500                1530   \n",
       "6967922  ENST00000707119                  1695                1725   \n",
       "6967923  ENST00000707119                  1710                1740   \n",
       "6968522  ENST00000707127                   735                 765   \n",
       "\n",
       "         difference  synapse_probability external_gene_name  \\\n",
       "6271       0.000000             0.014038         RANBP9-201   \n",
       "8725       0.000793             0.016235         MAP4K5-201   \n",
       "19034     -0.000488             0.014038        FAM168A-201   \n",
       "43148      0.000000             0.013428         SNAP91-201   \n",
       "43149      0.000000             0.013428         SNAP91-201   \n",
       "...             ...                  ...                ...   \n",
       "6964154    0.000244             0.016357           MADD-227   \n",
       "6965958    0.000305             0.009705         INPP4A-210   \n",
       "6967922   -0.000244             0.013611         ABLIM1-215   \n",
       "6967923   -0.000427             0.013428         ABLIM1-215   \n",
       "6968522    0.000000             0.015076          MEF2C-257   \n",
       "\n",
       "                                                                                      coords  \\\n",
       "6271              chr6:13644726-13644729:-,chr6:13652658-13652681:-,chr6:13657108-13657112:-   \n",
       "8725           chr14:50437995-50438008:-,chr14:50438091-50438094:-,chr14:50440012-50440026:-   \n",
       "19034          chr11:73419997-73419999:-,chr11:73425021-73425048:-,chr11:73430689-73430690:-   \n",
       "43148             chr6:83610675-83610677:-,chr6:83614856-83614862:-,chr6:83616968-83616990:-   \n",
       "43149             chr6:83610660-83610677:-,chr6:83614856-83614862:-,chr6:83616968-83616975:-   \n",
       "...                                                                                      ...   \n",
       "6964154        chr11:47324565-47324577:+,chr11:47326543-47326552:+,chr11:47326737-47326746:+   \n",
       "6965958           chr2:98546687-98546694:+,chr2:98548951-98548969:+,chr2:98552785-98552790:+   \n",
       "6967922  chr10:114439245-114439250:-,chr10:114440081-114440089:-,chr10:114441016-114441033:-   \n",
       "6967923  chr10:114439230-114439250:-,chr10:114440081-114440089:-,chr10:114441016-114441018:-   \n",
       "6968522           chr5:88729344-88729347:-,chr5:88730210-88730234:-,chr5:88731728-88731731:-   \n",
       "\n",
       "         num_exons  \n",
       "6271             3  \n",
       "8725             3  \n",
       "19034            3  \n",
       "43148            3  \n",
       "43149            3  \n",
       "...            ...  \n",
       "6964154          3  \n",
       "6965958          3  \n",
       "6967922          3  \n",
       "6967923          3  \n",
       "6968522          3  \n",
       "\n",
       "[1167 rows x 8 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged['num_exons'] = merged['coords'].apply(lambda x: x.count(',')+1)\n",
    "merged[merged['num_exons'] > 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ebf1005-6cc0-477f-9d49-427afbd85c24",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
